core/src/main/java/org/infinispan/remoting/inboundhandler/GlobalInboundInvocationHandler.java

```diff
@@ -154,14 +145,13 @@ private ReplicableCommandRunner(ReplicableCommand command, Reply reply,
       public void run() {
          try {
             CompletionStage<?> stage;
-            if (command instanceof GlobalRpcCommand) {
-               stage = ((GlobalRpcCommand) command).invokeAsync(globalComponentRegistry).whenComplete(this);
+            if (command instanceof GlobalRpcCommand globalRpcCommand) {
+               stage = globalRpcCommand.invokeAsync(globalComponentRegistry).whenComplete(this);
+               if (preserveOrder) {
+                  CompletionStages.join(stage);
```

Comment:
This is a bit worrisome to me.. I know it was like this before, but can we document how the blocking portion is handled here?

Reply:
> @wburns regular message (i.e. ordered) are delivered in a single thread per sender. We have to block the request until it is processed before handling the next one otherwise ordering can't be maintained.

That is what I figured, so in this case we are just blocking the JGroups thread then.